src/commands/serve.rs

```diff
@@ -0,0 +1,201 @@
+
+        let mut config = self.common.config(None)?;
+        config.wasm_component_model(true);
+        config.async_support(true);
+
+        let engine = Arc::new(Engine::new(&config)?);
+        let mut linker = Linker::new(&engine);
+
+        self.add_to_linker(&mut linker)?;
+
+        let component = Component::from_file(&engine, &self.component)?;
+
+        let instance = Arc::new(linker.instantiate_pre(&component)?);
+
+        let listener = tokio::net::TcpListener::bind(self.addr()).await?;
+
+        loop {
+            let (stream, _) = listener.accept().await?;
+            let engine = Arc::clone(&engine);
+            let instance = Arc::clone(&instance);
+            tokio::task::spawn(async move {
+                let handler = ProxyHandler::new(engine, instance);
+                if let Err(e) = http1::Builder::new()
+                    .keep_alive(true)
+                    .serve_connection(stream, handler)
+                    .await
+                {
+                    eprintln!("error: {e:?}");
+                }
+
+                Ok::<_, anyhow::Error>(())
+            });
+        }
+    }
+}
+
+#[derive(Clone)]
+struct ProxyHandler {
+    engine: Arc<Engine>,
+    instance_pre: Arc<InstancePre<Host>>,
+}
+
+impl ProxyHandler {
+    fn new(engine: Arc<Engine>, instance_pre: Arc<InstancePre<Host>>) -> Self {
+        Self {
+            engine,
+            instance_pre,
+        }
+    }
+}
+
+type Request = hyper::Request<hyper::body::Incoming>;
+
+impl hyper::service::Service<Request> for ProxyHandler {
+    type Response = hyper::Response<HyperOutgoingBody>;
+    type Error = anyhow::Error;
+    type Future = Pin<Box<dyn std::future::Future<Output = Result<Self::Response>> + Send>>;
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        use http_body_util::BodyExt;
+
+        let handler = self.clone();
+
+        Box::pin(async move {
+            let (sender, receiver) = tokio::sync::oneshot::channel();
+
+            // TODO: need to track the join handle, but don't want to block the response on it
+            tokio::task::spawn(async move {
+                let host = Host::new()?;
+                let mut store = Store::new(&handler.engine, host);
+
+                let req = store.data_mut().new_incoming_request(
+                    req.map(|body| body.map_err(|e| anyhow::anyhow!(e)).boxed()),
+                )?;
+
+                let out = store.data_mut().new_response_outparam(sender)?;
+
+                let (proxy, _inst) = wasmtime_wasi_http::proxy::Proxy::instantiate_pre(
+                    &mut store,
+                    &handler.instance_pre,
+                )
+                .await?;
+
+                proxy
+                    .wasi_http_incoming_handler()
+                    .call_handle(store, req, out)
+                    .await?;
+
+                Ok::<_, anyhow::Error>(())
+            });
+
+            let resp = receiver.await.unwrap()?;
+            Ok(resp)
```

Comment:
This can all be moved outside of the enclosing `async move` I think and the `Future` return could actually be just the receiver (or a boxed receiver, either way)

Reply:
Great idea! I like that this makes it so much more clear that we spawn the task and return the future that waits on the oneshot channel, rather than returning a future that does both.